src/Event.cs

```diff
@@ -62,41 +62,118 @@ public SubscriptionBuilder<EventContext> OnError(Callback callback)
+        private readonly SubscriptionBuilder<EventContext>.Callback? onError;
+        private SubscriptionBuilder<EventContext>.Callback? onEnded;
+
+        public bool IsEnded
+        {
+            get; private set;
+        }
+
+        public bool IsActive
+        {
+            get; private set;
+        }
 
         void ISubscriptionHandle.OnApplied(IEventContext ctx)
         {
             IsActive = true;
             onApplied?.Invoke((EventContext)ctx);
         }
 
+        void ISubscriptionHandle.OnEnded(IEventContext ctx)
+        {
+            IsActive = false;
+            IsEnded = true;
+            onEnded?.Invoke((EventContext)ctx);
+        }
+
+        void ISubscriptionHandle.OnError(IEventContext ctx)
+        {
+            IsActive = false;
+            IsEnded = true;
+            onError?.Invoke((EventContext)ctx);
+        }
+
+        /// <summary>
+        /// TODO: remove this constructor once legacy subscriptions are removed.
+        /// </summary>
+        /// <param name="conn"></param>
+        /// <param name="onApplied"></param>
+        /// <param name="onError"></param>
+        /// <param name="querySqls"></param>
         internal SubscriptionHandle(IDbConnection conn, SubscriptionBuilder<EventContext>.Callback? onApplied, SubscriptionBuilder<EventContext>.Callback? onError, string[] querySqls)
         {
+            IsActive = false;
+            IsEnded = false;
+            this.onApplied = onApplied;
+            conn.LegacySubscribe(this, querySqls);
+            // drop onError, as it will never be invoked for legacy subscriptions.
+        }
+
+        /// <summary>
+        /// Construct a subscription handle.
+        /// </summary>
+        /// <param name="conn"></param>
+        /// <param name="onApplied"></param>
+        /// <param name="onError"></param>
+        /// <param name="querySql"></param>
+        internal SubscriptionHandle(IDbConnection conn, SubscriptionBuilder<EventContext>.Callback? onApplied, SubscriptionBuilder<EventContext>.Callback? onError, string querySql)
+        {
+            IsActive = false;
+            IsEnded = false;
             this.onApplied = onApplied;
-            conn.Subscribe(this, querySqls);
+            this.onError = onError;
+            conn.Subscribe(this, querySql);
+        }
+
+        /// <summary>
+        /// Unsubscribe from the query controlled by this subscription handle.
+        /// 
+        /// Calling this more than once will result in an exception.
+        /// </summary>
+        public void Unsubscribe()
+        {
+            UnsubscribeThen(null);
+        }
+
+        /// <summary>
+        /// Unsubscribe from the query controlled by this subscription handle,
+        /// and call onEnded when its rows are removed from the client cache.
+        /// </summary>
+        public void UnsubscribeThen(SubscriptionBuilder<EventContext>.Callback? onEnded)
```

Comment:
Is implementation of this method supposed to actually unsubscribe by sending a websocket message or something?